### ISSUE TITLE
[Question] - Zero downtime deployment support in az cli (no yaml)

### ISSUE THREAD

**Issue by ilmax** [microsoft/azure-container-apps#244]
I'm trying to implement zero downtime deployment but I'm having a bit of an hard time, so I decided to open an issue here since I can't find what I'm looking for in the documentation.

I'm deploying a simple two container sample, the full source code is [here](https://github.com/ilmax/container-apps-sample) but I'm just focusing on the asp.net core part (Sample.Producer) for now.

My expectations are that in a single revision mode, before switching the ingress to the new revision, the health probes are exercised (at least **Readiness** and **Startup**). 

According to my experiments this doesn't seems to be the case, the previous active revision is deactivated while the new one is still provisioning, not waiting for it to be up & running. The portal also exposes it in the revision management pane as shown below:

![image](https://user-images.githubusercontent.com/3263338/171152801-c4620765-3da9-479c-abea-0b83b2d21134.png)

My next attempt was to use multiple active revisions as the same time, so essentially after pushing a new image to acr, I'm executing:
`az containerapp update -n {name} -g {rg} -i {new-image-just-pushed-to-acr}`

This is again switching the ingress to the provisioning revision, **unless the active revision is not the latest revision**, in this case the traffic is not switched to the newly created revision while it's provisioning, if before deployment the active revision is the latest, then deploying a new revision will switch the traffic to it while it's still in provisioning state.

I find this a bit strange, that's why I decided to open the following issue to ask for clarification.

Searching around the internet I've found that @denniszielke implemented zero downtime (or blue/green) with a yaml templates and some bash gymnastics (kudos to you Dennis), but it requires ~500 lines of code (see it [here](https://github.com/denniszielke/blue-green-with-containerapps/blob/main/scripts/deploy.sh)), seems a bit complicated and would love to have support in the az cli.

Am I missing something or the azure cli extension doesn't allow us to have zero downtime deployments without resorting to yaml files?

**Comment by ahmelsayed** [comment]
@ilmax You're right. Automatic switching doesn't wait for probes to pass, it'll immediately switch traffic and then the traffic will queue until the probes succeed or timeout. The only way to do zero-downtime/zero-coldstart deployment today is to use multiple revision mode with manual traffic shifting as you've described.

**Comment by ilmax** [comment]
Thank you for the answer @ahmelsayed. Is there any plan to improve it? I was expecting, given that the platform is built on top of k8s, that this functionality was built-in.

Implementing our own swap mechanism is doable but less than ideal, not to mention that the app environment may be internal so not even be exposed to the public internet.

If it's not on the backlog already, can we consider this as a feature request?

**Comment by anrub** [comment]
I think it would be useful, either with built-in functionality or by documenting some easy to use copy-and-paste for custom scripts.

We have the same problem with appservice, that there is no auto-swap in linux apps - for this function we built a custom github action, that contains a short script to wait for a healthy slot and then swaps. (so not every project needs to built this for themselfs). 

(blue-green deployment with auto-switching if health succeeds is one of the first questions I get from developers, following the max. instance limits for RAM/CPU)

**Comment by denniszielke** [comment]
Is there an update on this. We have talked to multiple customers that would expect seamless, reliable and automatic upgrades to happen in single revision mode.

**Comment by DanielMeixner** [comment]
Same here - my customer is running into the same issue, also having the expectation that probes are being honored before switching revisions. It would be great if this would come as an improvement. I also think this is what everyone is just expecting as a default behaviour.

**Comment by BigMorty** [comment]
This is on our backlog but no ETA at this time.

**Comment by timoknapp** [comment]
> This is on our backlog but no ETA at this time.

Hi @BigMorty , thanks! Could you maybe keep this thread updated in chase you have an ETA?

**Comment by BigMorty** [comment]
Yes, I will do my best.

**Comment by wesselkranenborg** [comment]
> Is there an update on this. We have talked to multiple customers that would expect seamless, reliable and automatic upgrades to happen in single revision mode.

Same here indeed, we also expected that and we really want to have zero-downtime deployments, just like you can 'easily' do with AKS (k8s).

**Comment by onionhammer** [comment]
Any update on this? Seems like it's blocking a lot of folks.

**Comment by denniszielke** [comment]
The second one can be removed by a simple ticket.
However I totally get your issue with the first one. Afaik it is work in progress.

**Comment by onionhammer** [comment]
@denniszielke easier said than done. I've had a ticket open since Friday and have had 0 movement other than some emails back and fourth.

**Comment by BigMorty** [comment]
I am happy to be able to say this is ON our Roadmap!

**Comment by onionhammer** [comment]
Here's a script I put together a few days ago to deploy a revision to ACA without taking down the current instance; it just juggles the traffic around, waits for healthcheck to succeed, then changes the traffic around again.

https://gist.github.com/onionhammer/df3f4af9d0e615b11ce062020508b15b

**Comment by ilmax** [comment]
@onionhammer Did you try to run some load testing while deploying? I find that a similar approach, was still causing downtime. My educated guess was that during the execution of `az containerapp ingress traffic set` there's a chance of causing downtime.

**Comment by onionhammer** [comment]
@ilmax Not load testing per se, just continually spamming chrome refresh on a page (with network logging persistent) I can see basically no difference when the page responds with a different revision name.

The trick for me was to have the application's startup probe do a lot of  the initialization work that subsequent requests may be slow to activate; for instance connecting to cosmos is very slow so the startup probe should be sure to hit cosmos so subsequent requests will have that cached.

**Comment by anrub** [comment]
@onionhammer @ilmax I also polled the response-code from the new revision (via curl in a loop), so that I do only switch to this new rev. when the status code is 200. For me, it was not enough to just wait for the deployment-cli to complete. It is also important to first activate the new revision (100% traffic) and only after that worked (with code 200) deactivate the previously active revision.

**Comment by onionhammer** [comment]
@ilmax The first step is important; finding which revision currently has 100% traffic and locking the traffic weighting rules to be revision based on that name before creating any new revisions.

![image](https://user-images.githubusercontent.com/969938/194139125-0d33845a-cdc6-4537-9ea9-53a98d87f059.png)

That should cause no downtime, then copying the revision should not affect the traffic, ping that new revision's specific URL until it's ready and then switch traffic again.

**Comment by onionhammer** [comment]
Sure, you could do that if you want some users to randomly sometimes get really long wait times.

**Comment by BigMorty** [comment]
Why is that. I am saying put say 5% to the new revision to make sure it is FULLY warmed up before moving everyone there.

**Comment by santi** [comment]
@BigMorty The problem is not that the application requires some requests to warmup, it is that the application has not been able to start at all, as the traffic switches over to the new revision even before the startup or readiness probes have successfully returned a response.

If the application uses 10 seconds to start before it can handle any requests (including startup / readiness probes), putting 5% of traffic to your new revision would cause 5% of your users to get a 10 seconds wait time.

Actually, the wait time would be even longer, as traffic is shifted as soon as the container is _provisioned_, meaning that not even the OS have had time to start.

**Comment by anrub** [comment]
> I am happy to be able to say this is ON our Roadmap!

Great, I would be great, if you could also keep the terraform use case in mind, so that I just can deploy another image version (an updated app deployment yaml) and the API juggles the zero-downtime update automatically, by respecting the health probes.

**Comment by BigMorty** [comment]
Thanks @santi, I understand the issue the platform is not solving, I was just adding on to how you could use @onionhammer's solution above.

**Comment by santi** [comment]
@BigMorty great! 😄 Currently working around this issue with some custom `az` CLI scripts in addition to `bicep` templates and a Python script, but looking forward to the feature release ⭐

**Comment by trzepak1916** [comment]
@onionhammer I'm using a similar solution as you do, but my observation is that
`az containerapp ingress traffic set --name <container-app-name> --resource-group <resource-group-name> --revision-weight <old_revision_name>=0 <new_revision_name>=100` takes some time to take effect and results in delayed responses from the container app (around 3-5 seconds), although I'm waiting for the new revision to become healthy and fully responsive before I switch the traffic to it.

I was trying both to curl application URL and monitor new revision's healthState before switching traffic to it, but both result in the same observation.

**Comment by onionhammer** [comment]
@trzepak1916 Interesting. I have not observed this, and I've tested it quite a bit (as I'm sure you have). 

Maybe a MSFTer can confirm whether when `az containerapp ingress traffic set` there is some delay before it becomes effective?

I don't know how precisely your script works, but during mine the chrome log looks like this; the highlighted row is where the new revision is being hit; I'm just hitting refreshing once a second - sorry it's not more sophisticated

![image](https://user-images.githubusercontent.com/969938/194576762-aaf865d2-7346-462c-b933-0284b2713dc8.png)

**Comment by ilmax** [comment]
@onionhammer did you tried load testing? Because what @trzepak1916 says is exactly what I was experiencing

**Comment by onionhammer** [comment]
@ilmax I will try it if you recommend a tool to use. Are you using AB or wrk or something?

**Comment by ilmax** [comment]
I was using wrk https://github.com/wg/wrk you can also use one of the many alternatives e.g. bombardier https://github.com/codesenberg/bombardier.

I was doing some experiment to see if ACA was a suitable alternative to AKS and one of the things I wanted was zero downtime deployment, that's were I found the issue.

I was simply returning ah hard-coded value from my endpoint and initializing the revision before setting it to activate.
Despite that I was still getting non 200 responses.

**Comment by onionhammer** [comment]
This is only my QA, which is scaled to a single 0.25 CPU instance, so not high performance for running a benchmark against.

10 seconds, without any deployments going on
```
  1 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    22.17ms   22.87ms 285.35ms   82.56%
    Req/Sec   610.06    230.43     1.16k    70.00%
  Latency Distribution
     50%    9.92ms
     75%   32.42ms
     90%   57.81ms
     99%   79.75ms
  3036 requests in 5.00s, 1.24MB read
Requests/sec:    606.75
Transfer/sec:    253.35KB
```

This is 90 seconds while a deployment is going on (i.e. running my powershell script above) - I ran this twice

```
  1 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    22.47ms   30.85ms 492.26ms   87.45%
    Req/Sec   734.68    352.06     1.28k    59.56%
  Latency Distribution
     50%    8.73ms
     75%   23.62ms
     90%   62.45ms
     99%  163.32ms
  65850 requests in 1.50m, 26.30MB read
Requests/sec:    731.21
Transfer/sec:    299.10KB
```

**Comment by ahmelsayed** [comment]
The default behavior now should be that `latestRevision: true` weight should be going to the latestReadyRevision, rather than latestRevision. So you shouldn't need any custom scripts to get zero downtime deployment working with both single and multiple revision modes. 

The api version 2022-10-01 (still pending release) will add a readonly property `latestReadyRevision` to indicate which revision the system considers to be the latest ready (vs latest created). A revision becomes ready once it passes its probes at least once.

**Comment by onionhammer** [comment]
Hi @ahmelsayed that sounds exciting `-` can you break that down a bit more, in practical terms, what that means for us?

Like for bicep files, would we use `latestReadyRevision` now? What does this mean for the CLI? When can we utilize these changes?

**Comment by ahmelsayed** [comment]
There is a property on the container app called `latestRevisionName`. The value in that property is the name of latest _created_ revision. It was also the same revision referred to in the ingress section when you use `latestRevision: true`. e.g: the default

```json
{
  "latestRevision": true,
  "weight": 100
}
```

`latestRevisionName` still is the latest _created_ revision name. There will be a new property `latestReadyRevisionName` (once 2022-10-01 is available). `latestRevision: true` in traffic now refers to the `latestReadyRevisionName` instead of `latestRevisionName` (_aka: latest created revision_)


The latest **created** revision becomes the latest **ready** revision once it passes its probes at least once. 

For example, if you have an app `foo` with a first revision being created on app creation `foo--rev1`:

1. `latestRevisionName == foo--rev1` and `latestReadyRevisionName == foo--rev1` _Since `rev1` this is the first/only revision created_. 

2. Create `foo--rev2`:  
`latestRevisionName == foo--rev2`, but `latestReadyRevisionName == foo--rev1`. While waiting for `--rev2` to pass its probes. 

3. if:
    * `rev2` never passes its probes, it never becomes `latestReadyRevisionName` and never gets traffic. 
    * `rev2` passes its probes 2 minutes later, it becomes `latestReadyRevisionName` 2 minutes later and traffic goes to it. 
 

### **TL;DR:**
When you assign traffic to `latestRevision: true` (_which is the default of 100% to `latestRevision: true`_), you're assigning traffic to the **latest created revision that has passed its probes at least once**. As opposed to just the latest created revision (as it used to be). 

**In bicep:** `latestRevisionName` still has the same semantics it always had (i.e: the name of the latest created revision) with 2022-10-01, there will be the new `latestReadyRevisionName`. Depending on what you're doing in bicep, you can use either.

**Comment by ahmelsayed** [comment]
Here is an example where I create an app with an initial revision `--rev1`, then start a load test and create another 4 revisions. The result are similar to yours without the need for a custom script to orchestrate traffic updates. 

![WindowsTerminal_2XMi2EGqva](https://user-images.githubusercontent.com/645740/200914558-494882ed-3f81-45c9-bfd9-e5bd369c5744.gif)

**Comment by onionhammer** [comment]
Thank you for the explanation, @ahmelsayed

It sounds like the default behavior (once 2022-10-01 comes out) of setting traffic to `latestRevision: true` will be zero-downtime deployments.

**Comment by ahmelsayed** [comment]
> It sounds like the default behavior of setting traffic to latestRevision: true will be zero-downtime deployments - still unclear if that has already taken effect or not

Yes, this has taken effect as of last week of October/first week of November (depending on the region and the rollout of the update).

> Is this true for single revision mode as well as multi-revision mode?

Yes, this works for both single and multi-revision.

**Comment by ahmelsayed** [comment]
I'll close this issue too as complete, and track CLI updates to allow waiting for the traffic to have shifted in https://github.com/microsoft/azure-container-apps/issues/487

**Comment by anrub** [comment]
@ahmelsayed Thanks for adding that feature, but I have to add, that it is now not possible, to check if the deployment was successful or not. So my pipeline will always deploy successful, although the app maybe did not start healthy. Is it planned, to realize that it is possbile to know, if the deployment was successful?

**Comment by hamenu** [comment]
Hey @ahmelsayed, I still experience this issue in a _single-revision_ container app with just one container, even after setting up the probes.

My app doesn't have separate endpoints for startup/readiness/liveness, so I use a single `/version` endpoint for all of them:

```
STARTUP
  path: /version
  timeoutSeconds: 15
  periodSeconds: 10
  successThreshold: 1
  failureThreshold: 3

READINESS
  path: /version
  timeoutSeconds: 10
  periodSeconds: 5
  successThreshold: 1
  failureThreshold: 3

LIVENESS
  path: /version
  timeoutSeconds: 5
  periodSeconds: 3
  successThreshold: 1
  failureThreshold: 3
```

The endpoint is very light and just returns assembly version, commit info, but more importantly: the value of `CONTAINER_APP_REVISION` environment variable.

If I use Azure Portal to create a new revision, this is the outcome:

![image](https://user-images.githubusercontent.com/123523199/214474027-e196ae3a-f405-4ae4-a17e-8e170ea9d0c6.png)

I have tried selecting/unselecting checkbox no. 2, no difference.

![image](https://user-images.githubusercontent.com/123523199/214474120-daf28524-2df9-496c-bfe7-add7c779409e.png)

I also use Azure CLI in the pipeline where I update just the image:

```
latest_revision_name=$(az containerapp update --resource-group <resource-group> --name <container-app-name> --image <image-from-azure-container-registry> --query 'properties.latestRevisionName' --output tsv)

az containerapp revision show --resource-group <resource-group> --name <container-app-name> --revision $latest_revision_name --query '{ active: properties.active, healthState: properties.healthState, provisioningState: properties.provisioningState, trafficWeight: properties.trafficWeight }'
```

This is the output:

```
{
  "active": true, // [x] unexpected when healthState=None and provisioningState=Provisioning
  "healthState": "None",
  "provisioningState": "Provisioning",       
  "trafficWeight": 100 // [x] unexpected when healthState=None and provisioningState=Provisioning
}
```

During this period (Provision**ing**), the `/version` endpoint stops returning a response and the spinner in the browser keeps spinning.

Once the new revision becomes Provision**ed**, the `/version` endpoint returns the new revision name, but it experienced 5 seconds of no response - not sure if that's considered as downtime or cold start.

---
Relates to a new issue I have created: https://github.com/microsoft/azure-container-apps/issues/598
